File: frontend/__tests__/fixtures/inventory.ts
import type { InventoryLiveResponse } from "@/types/api";

/**
 * Deterministic fixture mimicking GET /api/inventory — two years of
 * weekly bars with a gentle downward drift.
 */
export function makeInventoryFixture(weeks = 104): InventoryLiveResponse {
  const start = new Date("2024-04-24T00:00:00Z");
  const history = Array.from({ length: weeks }, (_, i) => {
    const d = new Date(start);
    d.setUTCDate(d.getUTCDate() + i * 7);
    const commercial = 430_000_000 - i * 250_000 + Math.sin(i / 6) * 3_000_000;
    const spr = 370_000_000 + Math.cos(i / 9) * 1_000_000;
    const cushing = 45_000_000 + Math.sin(i / 4) * 4_000_000;
    return {
      date: d.toISOString().slice(0, 10),
      commercial_bbls: commercial,
      spr_bbls: spr,
      cushing_bbls: cushing,
      total_bbls: commercial + spr,
    };
  });
  const latest = history[history.length - 1]!;
  return {
    commercial_bbls: latest.commercial_bbls ?? 400_000_000,
    spr_bbls: latest.spr_bbls ?? 370_000_000,
    cushing_bbls: latest.cushing_bbls ?? 45_000_000,
    total_bbls: latest.total_bbls ?? 770_000_000,
    as_of: latest.date,
    source: "EIA",
    history,
    forecast: {
      daily_depletion_bbls: -36_000,
      weekly_depletion_bbls: -252_000,
      projected_floor_date: "2026-11-15",
      r_squared: 0.72,
      floor_bbls: 360_000_000,
    },
  };
}

File: backend/models/inventory.py
"""Inventory response schema.

Shapes the output of ``providers.inventory.fetch_inventory`` +
``quantitative_models.forecast_depletion`` into a JSON payload with
current stocks, 2y history, and a simple linear-regression depletion
forecast.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class InventoryPoint(BaseModel):
    """One weekly observation."""

    date: str = Field(..., description="ISO-8601 date of the EIA release")
    commercial_bbls: float | None = Field(
        default=None, description="U.S. commercial crude stocks (bbls)"
    )
    spr_bbls: float | None = Field(
        default=None, description="Strategic Petroleum Reserve stocks (bbls)"
    )
    cushing_bbls: float | None = Field(
        default=None, description="Cushing, OK hub stocks (bbls)"
    )
    total_bbls: float | None = Field(
        default=None, description="Commercial + SPR (bbls)"
    )


class DepletionForecast(BaseModel):
    """Linear-regression forecast from ``forecast_depletion``."""

    daily_depletion_bbls: float = Field(
        ..., description="Slope in bbls/day (negative = drawdown)"
    )
    weekly_depletion_bbls: float = Field(
        ..., description="Slope * 7 (bbls/week)"
    )
    projected_floor_date: str | None = Field(
        default=None,
        description="ISO-8601 date when total breaches floor_bbls, or null",
    )
    r_squared: float = Field(
        ..., description="R^2 of the linear fit on the trailing window"
    )
    floor_bbls: float = Field(..., description="Configured floor threshold (bbls)")


class InventoryResponse(BaseModel):
    """Current stocks + 2y history + depletion forecast."""

    commercial_bbls: float = Field(..., description="Latest commercial (bbls)")
    spr_bbls: float = Field(..., description="Latest SPR (bbls)")
    cushing_bbls: float = Field(..., description="Latest Cushing (bbls)")
    total_bbls: float = Field(..., description="Latest commercial + SPR (bbls)")
    as_of: str = Field(..., description="ISO-8601 date of the latest release")
    source: str = Field(..., description="Upstream provider tag (e.g. 'EIA')")
    history: list[InventoryPoint] = Field(
        default_factory=list,
        description="Trailing 2y of weekly observations (ascending).",
    )
    forecast: DepletionForecast = Field(..., description="Linear depletion fit")
